File: libraries/python/src/urc_converter/parser/state.py
"""Mutable scan state shared by the parser modules."""

from dataclasses import dataclass, field

from ..model import TimingPoint, Version
from ..strings import SECTION_URC


@dataclass
class RawNote:
    timestamp: int
    lane: int
    type_token: str
    line: int


@dataclass
class ParseState:
    seen: set[str] = field(default_factory=lambda: {SECTION_URC})
    last_index: int = 0
    version: Version = field(default_factory=lambda: Version(1, 1))
    metadata: dict[str, str] = field(default_factory=dict)
    windows: list[float] | None = None
    rates: list[float] | None = None
    layout_type: tuple[int, int] | None = None
    special: list[int] | None = None
    special_seen: bool = False
    timing: list[TimingPoint] = field(default_factory=list)
    notes: list[RawNote] = field(default_factory=list)

File: libraries/dotnet/UrcConverter/Parser/State.fs
namespace UrcConverter.Parser

open UrcConverter

module internal State =

    [<RequireQualifiedAccess>]
    type Section =
        | UrcHeader
        | Metadata
        | Judgment
        | Layout
        | Timing
        | Notes

    let private allSections =
        [
            Section.UrcHeader
            Section.Metadata
            Section.Judgment
            Section.Layout
            Section.Timing
            Section.Notes
        ]

    let byName =
        List.zip Strings.sections allSections
        |> List.indexed
        |> List.map (fun (index, (name, section)) -> name, (section, index))
        |> Map.ofList

    type RawNote =
        {
            Timestamp: int
            Lane: int
            TypeToken: string
            Line: int
        }

    // TimingPoints and Notes accumulate reversed; build reverses them back.
    type ParseState =
        {
            Seen: Set<string>
            LastIndex: int
            Version: Version
            Metadata: (string * string) list
            Windows: float list option
            Rates: float list option
            LayoutType: (int * int) option
            Special: int list option
            SpecialSeen: bool
            TimingPoints: TimingPoint list
            Notes: RawNote list
        }

    let initial =
        {
            Seen = Set.ofList [ Strings.sectionUrc ]
            LastIndex = 0
            Version = { Major = 1; Minor = 1 }
            Metadata = []
            Windows = None
            Rates = None
            LayoutType = None
            Special = None
            SpecialSeen = false
            TimingPoints = []
            Notes = []
        }
